mempool/cat/reactor.go

```diff
@@ -0,0 +1,446 @@
+	case *protomem.SeenTx:
+		txKey, err := types.TxKeyFromBytes(msg.TxKey)
+		if err != nil {
+			memR.Logger.Error("peer sent SeenTx with incorrect tx key", "err", err)
+			memR.Switch.StopPeerForError(e.Src, err)
+			return
+		}
+		peerID := memR.ids.GetIDForPeer(e.Src.ID())
+		memR.mempool.PeerHasTx(peerID, txKey)
+		// Check if we don't already have the transaction and that it was recently rejected
+		if !memR.mempool.Has(txKey) && !memR.mempool.IsRejectedTx(txKey) {
+			// If we are already requesting that tx, then we don't need to go any further.
+			if memR.requests.ForTx(txKey) != 0 {
+				memR.Logger.Debug("received a SeenTx message for a transaction we are already requesting", "txKey", txKey)
+				return
+			}
+
+			// If it was a low-priority transaction and we don't have capacity, then ignore.
+			if memR.mempool.WasRecentlyEvicted(txKey) {
+				if !memR.mempool.CanFitEvictedTx(txKey) {
+					return
+				}
+			}
+
+			// We don't have the transaction, nor are we requesting it so we send the node
+			// a want msg
+			memR.requestTx(txKey, e.Src)
+		} else {
+			memR.Logger.Debug("received a seen tx for a tx we already have", "txKey", txKey)
+		}
+
+	// A peer is requesting a transaction that we have claimed to have. Find the specified
+	// transaction and broadcast it to the peer. We may no longer have the transaction
+	case *protomem.WantTx:
+		txKey, err := types.TxKeyFromBytes(msg.TxKey)
+		if err != nil {
+			memR.Logger.Error("peer sent WantTx with incorrect tx key", "err", err)
+			memR.Switch.StopPeerForError(e.Src, err)
+			return
+		}
+		tx, has := memR.mempool.Get(txKey)
+		if has && !memR.opts.ListenOnly {
+			peerID := memR.ids.GetIDForPeer(e.Src.ID())
+			memR.Logger.Debug("sending a tx in response to a want msg", "peer", peerID)
+			if p2p.SendEnvelopeShim(e.Src, p2p.Envelope{
+				ChannelID: mempool.MempoolChannel,
+				Message:   &protomem.Txs{Txs: [][]byte{tx}},
+			}, memR.Logger) {
+				memR.mempool.PeerHasTx(peerID, txKey)
+			}
+		}
+
+	default:
+		memR.Logger.Error("unknown message type", "src", e.Src, "chId", e.ChannelID, "msg", fmt.Sprintf("%T", msg))
+		memR.Switch.StopPeerForError(e.Src, fmt.Errorf("mempool cannot handle message of type: %T", msg))
+		return
+	}
+}
+
+// PeerState describes the state of a peer.
+type PeerState interface {
+	GetHeight() int64
+}
+
+// broadcastSeenTx broadcasts a SeenTx message to all peers unless we
+// know they have already seen the transaction
+func (memR *Reactor) broadcastSeenTx(txKey types.TxKey) {
+	memR.Logger.Debug("broadcasting seen tx to all peers", "tx_key", txKey.String())
+	msg := &protomem.Message{
+		Sum: &protomem.Message_SeenTx{
+			SeenTx: &protomem.SeenTx{
+				TxKey: txKey[:],
+			},
+		},
+	}
+	bz, err := msg.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	// Add jitter to when the node broadcasts it's seen txs to stagger when nodes
+	// in the network broadcast their seenTx messages.
+	time.Sleep(time.Duration(rand.Intn(10)*10) * time.Millisecond)
```

Comment:
Non-critical. We could remove it. Staggering when we send these messages may reduce the amount we need to send but it might mean in the worst case that a node becomes aware of transactions existence later than necessary.